teachertool/src/transforms/runEvaluateAsync.ts

```diff
@@ -1,18 +1,79 @@
-import { runEvalInEditorAsync } from "../services/makecodeEditorService";
+import { logError } from "../services/loggingService";
+import { runValidatorPlanAsync } from "../services/makecodeEditorService";
 import { stateAndDispatch } from "../state";
 import * as Actions from "../state/actions";
+import { getCatalogCriteriaWithId } from "../state/helpers";
+import { CriteriaEvaluationResult, CriteriaInstance } from "../types/criteria";
+import { ErrorCode } from "../types/errorCode";
 import { makeNotification } from "../utils";
 import { postNotification } from "./postNotification";
 
-export async function runEvaluateAsync(rubric: string) {
-    const { dispatch } = stateAndDispatch();
+function generateValidatorPlan(criteriaInstance: CriteriaInstance): pxt.blocks.ValidatorPlan | undefined {
+    const { state: teacherTool } = stateAndDispatch();
 
-    const evalResult = await runEvalInEditorAsync(rubric);
-    if (evalResult) {
-        dispatch(Actions.setEvalResult(evalResult));
-    } else {
-        // Errors already logged in the runEvalInEditorAsync function.
-        // Just notify the user.
-        postNotification(makeNotification(lf("Unable to evaluate project"), 2000));
+    const catalogCriteria = getCatalogCriteriaWithId(criteriaInstance.catalogCriteriaId);
+    if (!catalogCriteria) {
+        logError(ErrorCode.evalMissingCriteria, "Attempting to evaluate criteria with unrecognized catalog id", {
+            id: criteriaInstance.catalogCriteriaId,
+        });
+        return undefined;
+    }
+
+    const plan = teacherTool.validatorPlans?.find(plan => plan.name === catalogCriteria.use);
+    if (!plan) {
+        logError(ErrorCode.evalMissingPlan, "Attempting to evaluate criteria with unrecognized plan", {
+            plan: catalogCriteria.use,
+        });
+        return undefined;
+    }
+
+    // TODO: Fill in any parameters. Error if parameters are missing.
+
+    return plan;
+}
+
+export async function runEvaluateAsync() {
+    const { state: teacherTool, dispatch } = stateAndDispatch();
+
+    // Clear all existing results.
+    dispatch(Actions.clearAllEvalResults());
+
+    // EvalRequest promises will resolve to true if evaluation completed successfully (regarless of pass/fail)
+    // They will only resolve to false if evaluation was unable to complete.
+    const evalRequests = teacherTool.selectedCriteria.map(
+        criteriaInstance =>
+            new Promise(async resolve => {
+                dispatch(Actions.setEvalResult(criteriaInstance.instanceId, CriteriaEvaluationResult.InProgress));
+
+                const plan = generateValidatorPlan(criteriaInstance);
+
+                if (!plan) {
+                    dispatch(Actions.clearEvalResult(criteriaInstance.instanceId));
+                    resolve(false);
```

Comment:
It may not be necessary, but I like to include a `return` when resolving the promise just to make it really clear the code will not continue executing from this point.
```suggestion
                    return resolve(false);
```